### PR TITLE
213700: ADO Private Build agent

### PR DESCRIPTION
**Pull request by ajaysoni-cap**
# Description
This pr includes all the changes to set up a VM scale-set based ADO build agents.
Changes include,
- PowerShell script to deploy the scaleset
- Independent Pipeline to deploy the PowerShell script
- Add new API Permission permissions on app reg
- Attach route table to build agent subnet
- Add an NSG for the build agent (rules will need to be refined)

Link to the relevant work items: [AB#213700](https://dev.azure.com/defragovuk/93c65135-d16b-49d6-a191-4a5313532779/_workitems/edit/213700)

# Testing
- Verified the App Registration needed to read the images from shared gallery.
- Tested the pipeline to set up the agent pool.

### CODE REVIEW

**Review by DaRoszko — DISMISSED** [review]

**Review by ppotturi — APPROVED** [review]

File: .azuredevops/deploy-platform-core.yaml
[the remark this answers is not in the document]
Comment (ajaysoni-cap) [reply .azuredevops/deploy-platform-core.yaml:1]: not sure what you are referring to here?

[the remark this answers is not in the document]
Comment (rajeshkaremane) [reply .azuredevops/deploy-platform-core.yaml:1]: it is related to line 99 bootstrap pipeline

[the remark this answers is not in the document]
Comment (ajaysoni-cap) [reply .azuredevops/deploy-platform-core.yaml:1]: No, we agreed to keep it at both places....
Most likely you will run Bootstrap only once... but the network updates will be often

File: scripts/Create-AdoBuildAgent.ps1 [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        $output = az vmss create `
+            --resource-group $resourceGroup `
+            --name $vmssName `
+            --computer-name-prefix $vmssName `
+            --vm-sku Standard_D4s_v4 `
+            --instance-count 2 `
+            --subnet $subnetId `
+            --image "$imageId" `
+            --authentication-type password `
+            --admin-username $adoAgentUser `
```
Comment (rajeshkaremane) [inline scripts/Create-AdoBuildAgent.ps1:105]: Probably you will have to convert this back to normal string same for password

File: scripts/Create-AdoBuildAgent.ps1 [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    [Parameter(Mandatory)]
+    [string] $vmssName,
+    [Parameter(Mandatory)]
+    [string] $subnetId,
+    [Parameter(Mandatory)]
+    [string] $imageId,
+    [Parameter(Mandatory)]
+    [string] $adoAgentUser,
+    [Parameter(Mandatory)]
+    [string] $adoAgentPass
```
[the remark this answers is not in the document]
Comment (ajaysoni-cap) [reply scripts/Create-AdoBuildAgent.ps1:48]: as the values are coming from KV they are not printed in clear text..
also it throws a conversion error when parameter is defined as securestring https://dev.azure.com/defragovuk/DEFRA-FFC/_build/results?buildId=348998&view=logs&j=b3138091-3170-55f4-da31-f763ee2a89a0&t=4386c34f-f137-57cc-4fa9-e3b551d4d1ef&l=277

**Review by rajeshkaremane — APPROVED** [review]

**Review by ppotturi — APPROVED** [review]

**Pull request merged**